File: agent-runtime/src/ocr-budget.ts
/**
 * How much of an OCR'd document actually goes into the prompt.
 *
 * ## The decision this makes, and why it has to be made somewhere
 *
 * `compose_prompt_with_attachments` pastes every attachment's text into the
 * turn, whole. For a one-page invoice that is exactly right. For a 40-page
 * scanned standard it is the failure `context-ledger.ts` was written to
 * diagnose — "Evidence dominates: whole documents are reaching the window
 * instead of references" — except that arriving at the diagnosis after the fact
 * does not help the run that already lost its history to it.
 *
 * So the size decision happens before the paste, and it is stated as a rule
 * rather than left to a threshold buried in a call site.
 *
 * ## The rule
 *
 * Let
 *
 * - `window` be the model's context window,
 * - `committed` be everything already spoken for this turn — the system prompt,
 *   the tool schemas, the notes, the transcript so far, and the reserve held
 *   back for the reply,
 * - `budget = window - committed`, what is genuinely free, and
 * - `docTokens` be the document's text, counted.
 *
 * Then:
 *
 * - `docTokens <= FULL_INCLUSION_SHARE * budget` → **the whole text goes in.**
 * - otherwise → **partial inclusion**: as much of the document as the
 *   allowance holds is injected, and the turn says plainly how much that was.
 *   (Ranking those passages by relevance to the question is the natural next
 *   step here and is deliberately not claimed yet — see the note below.)
 *
 * ## Why the share is a half and not all of it
 *
 * A document permitted to fill the whole free budget leaves nothing for the
 * answer to build on, or for a second turn to exist in. The person asks a
 * follow-up and the run compacts immediately, losing the very document it just
 * read. Half leaves room for the conversation the document was attached *for*.
 *
 * The number is a judgement, not a measurement, and it is written here once so
 * that it can be argued with in one place.
 *
 * ## Why not summarisation
 *
 * Summarising is the third option and it was not taken. It costs a second model
 * call on a machine that may have no GPU left after the first, and it replaces
 * the document's own words with a paraphrase — in a system whose evidence is
 * meant to be quotable back to the page it came from. Partial inclusion keeps
 * every word it does include verbatim.
 *
 * ## What this does not yet do
 *
 * The part included is the beginning of the document, not the part most
 * relevant to the question. {@link fill} is written against *ranked* chunks so
 * that relevance ordering can be dropped in without changing the budget rule —
 * but until something actually ranks them, no message here claims it does.
 */

/** The share of free budget a single document's text may occupy. */
export const FULL_INCLUSION_SHARE = 0.5;

/**
 * The floor under a document's allowance, in tokens.
 *
 * Without it, a turn that is already nearly full computes a budget of a few
 * hundred tokens and admits a handful of sentences — which reads to the model
 * as a document that says almost nothing, rather than as a document that could
 * not be shown. Below this floor the honest outcome is to say it does not fit,
 * which is what `plan` returns.
 */
export const MINIMUM_USEFUL_ALLOWANCE = 512;

/** What to do with one document's text. */
export type InjectionStrategy = "full" | "chunked" | "reference-only";

export interface InjectionPlan {
  strategy: InjectionStrategy;
  /** Tokens this document may spend. */
  allowance: number;
  /** The document's full size, as counted. */
  documentTokens: number;
  /** Free budget at the moment of the decision. */
  budget: number;
  /**
   * Why, in the words the person reads. Shown verbatim on the context row —
   * "about 30% of it" is the difference between a trustworthy answer and one
   * that silently rests on a third of the evidence.
   */
  explanation: string;
}

/**
 * Decides how one document enters the prompt.
 *
 * `window` of zero means the runtime was never told the model's window. That is
 * reported as `full` with an explanation saying so, rather than as a refusal:
 * an unknown window is not evidence that the document does not fit, and
 * refusing on it would break every run against a model whose window nobody
 * configured. The overflow, if it comes, is then the provider's own clear error
 * rather than a guess made here.
 */
export function plan(input: {
  documentTokens: number;
  committed: number;
  window: number;
}): InjectionPlan {
  const { documentTokens, committed, window } = input;

  if (window <= 0) {
    return {
      strategy: "full",
      allowance: documentTokens,
      documentTokens,
      budget: 0,
      explanation:
        "The model's context window is not known to this runtime, so the whole document was included. If it does not fit, the model server will say so.",
    };
  }

  const budget = Math.max(0, window - committed);
  const allowance = Math.floor(budget * FULL_INCLUSION_SHARE);

  if (documentTokens <= allowance) {
    return {
      strategy: "full",
      allowance,
      documentTokens,
      budget,
      explanation: `The whole document was included — ${documentTokens.toLocaleString()} tokens, within the ${allowance.toLocaleString()} available to it.`,
    };
  }

  if (allowance < MINIMUM_USEFUL_ALLOWANCE) {
    return {
      strategy: "reference-only",
      allowance: 0,
      documentTokens,
      budget,
      explanation: `There was no room for this document — ${documentTokens.toLocaleString()} tokens against ${allowance.toLocaleString()} available. It was read, but none of it is in this turn.`,
    };
  }

  const share = Math.round((allowance / documentTokens) * 100);
  return {
    strategy: "chunked",
    allowance,
    documentTokens,
    budget,
    explanation: `This document is ${documentTokens.toLocaleString()} tokens and ${allowance.toLocaleString()} were available, so roughly the first ${share}% of it was included. The rest is not in this turn.`,
  };
}

/**
 * Fills an allowance from ranked chunks, in rank order.
 *
 * Stops at the first chunk that does not fit rather than skipping it to fit a
 * later, smaller one. Packing greedily would reorder the document by size, and
 * a model handed passage 9 and passage 3 but not passage 4 reads a text with a
 * hole in it that nothing marks — which is how a confident answer gets built on
 * a paragraph that was never there.
 */
export function fill<T extends { tokens: number }>(
  ranked: readonly T[],
  allowance: number,
): { taken: T[]; used: number; omitted: number } {
  const taken: T[] = [];
  let used = 0;
  let omitted = 0;
  let stopped = false;
  for (const chunk of ranked) {
    if (!stopped && used + chunk.tokens <= allowance) {
      taken.push(chunk);
      used += chunk.tokens;
    } else {
      stopped = true;
      omitted += 1;
    }
  }
  return { taken, used, omitted };
}

File: src-tauri/src/ai_engine/ocr_budget.rs
//! How much of an OCR'd document is allowed into the prompt.
//!
//! ## Why this exists here as well as in the runtime
//!
//! The rule is written twice, and that is a cost worth naming rather than
//! hiding. `agent-runtime/src/ocr-budget.ts` holds the same thresholds because
//! the runtime needs them to explain its own ledger. This copy is the one that
//! **decides**, because the prompt is composed on this side — in
//! `compose_prompt_with_attachments`, before the runtime is handed anything at
//! all. Putting the decision in the runtime would mean shipping the whole
//! 40-page document across the wire in order to be told it did not fit.
//!
//! The two are kept in step by the constants below being asserted against the
//! same boundary in both test suites. If they ever disagree, the Rust side is
//! what actually happened and the TypeScript side is what the screen claimed —
//! which is the sort of divergence this repository has been bitten by before.
//!
//! ## The rule
//!
//! With `budget = window - committed` (what is genuinely free this turn) and
//! `document` the OCR'd text's size:
//!
//! - `document <= FULL_INCLUSION_SHARE * budget` → the whole text goes in.
//! - otherwise, if the resulting allowance is at least
//!   [`MINIMUM_USEFUL_ALLOWANCE`] → as much of the document as that allowance
//!   holds is injected, and the turn says plainly how much that was.
//! - otherwise → nothing of it goes in, and the turn says so plainly.
//!
//! ## Why half, and not all of the free budget
//!
//! A document permitted to fill the whole of what is free leaves nothing for
//! the answer or for a second turn. The person asks a follow-up, the run
//! compacts immediately, and the first thing it loses is the document it was
//! just given. Half leaves room for the conversation the document was attached
//! for.
//!
//! ## Why not summarisation
//!
//! It costs a second model call on a machine that may have nothing left after
//! the first, and it replaces the document's own words with a paraphrase — in a
//! system whose evidence is meant to be quotable back to the page it came from.
//! Taking a prefix keeps every word it does include verbatim.
//!
//! ## What this does not yet do
//!
//! The part that is included is the *beginning* of the document, not the part
//! most relevant to the question. Ranking passages against the question — the
//! knowledge index already has the machinery, in `KnowledgeIndex::search` — is
//! the obvious improvement and is deliberately not claimed here until it is
//! built: `explanation` is shown to the person verbatim, and a sentence saying
//! the relevant passages were chosen, over a prefix that was not, would be a
//! lie about the evidence an answer rests on. [`take_tokens`] is the seam.

use serde::{Deserialize, Serialize};

/// The share of free budget one document's text may occupy.
pub const FULL_INCLUSION_SHARE: f64 = 0.5;

/// The floor under a document's allowance, in tokens.
///
/// Below this, injecting "the document" means injecting three sentences of it.
/// A model reading that answers confidently from a fragment nobody has marked
/// as a fragment, which is worse than being told the document did not fit.
pub const MINIMUM_USEFUL_ALLOWANCE: u32 = 512;

/// Characters per token, for the pre-call estimate.
///
/// Four is the usual figure for English prose and is wrong for dense tables,
/// which is exactly what OCR output often is. It is labelled an estimate
/// wherever it travels, and the reconciliation on the next model turn replaces
/// it with what was actually charged. See
/// [`crate::ai_engine::token_reconciliation`].
const CHARS_PER_TOKEN: usize = 4;

/// A character-count estimate of a document's token cost.
///
/// Deliberately not called `count`: nothing here counted anything.
pub fn estimate_tokens(text: &str) -> u32 {
    let chars = text.chars().count();
    if chars == 0 {
        return 0;
    }
    ((chars / CHARS_PER_TOKEN) as u32).max(1)
}

/// What will be done with one document's text.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum InjectionStrategy {
    /// The whole text goes into the prompt.
    Full,
    /// Only as much of the document as fits goes in; the rest is left out.
    Chunked,
    /// None of it fits. The document is named, nothing more.
    ReferenceOnly,
}

impl InjectionStrategy {
    pub fn label(self) -> &'static str {
        match self {
            InjectionStrategy::Full => "in full",
            InjectionStrategy::Chunked => "in part",
            InjectionStrategy::ReferenceOnly => "by reference only",
        }
    }
}

/// The decision, with the numbers it was made from.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InjectionPlan {
    pub strategy: InjectionStrategy,
    /// Tokens this document may spend.
    pub allowance: u32,
    /// The document's full estimated size.
    pub document_tokens: u32,
    /// What was free before this document was considered.
    pub budget: u32,
    /// Why, in the words the person reads. Shown verbatim.
    pub explanation: String,
}

/// Decides how one document enters the prompt.
///
/// A `window` of zero means nobody told this process the model's context size.
/// That yields `Full` with an explanation saying so, rather than a refusal: an
/// unknown window is not evidence that the document does not fit, and refusing
/// on it would break every run against an unconfigured model. If it really does
/// not fit, the runtime's own tokenizer check refuses with an exact figure —
/// which is a better error than a guess made here.
pub fn plan(document_tokens: u32, committed: u32, window: u32) -> InjectionPlan {
    if window == 0 {
        return InjectionPlan {
            strategy: InjectionStrategy::Full,
            allowance: document_tokens,
            document_tokens,
            budget: 0,
            explanation:
                "The model's context window is not known here, so the whole document was included."
                    .to_string(),
        };
    }

    let budget = window.saturating_sub(committed);
    let allowance = (f64::from(budget) * FULL_INCLUSION_SHARE) as u32;

    if document_tokens <= allowance {
        return InjectionPlan {
            strategy: InjectionStrategy::Full,
            allowance,
            document_tokens,
            budget,
            explanation: format!(
                "The whole document was included — about {document_tokens} tokens, within the {allowance} available to it."
            ),
        };
    }

    if allowance < MINIMUM_USEFUL_ALLOWANCE {
        return InjectionPlan {
            strategy: InjectionStrategy::ReferenceOnly,
            allowance: 0,
            document_tokens,
            budget,
            explanation: format!(
                "There was no room for this document — about {document_tokens} tokens against {allowance} available. \
                 It was read, but none of its text is in this turn."
            ),
        };
    }

    let share = ((f64::from(allowance) / f64::from(document_tokens)) * 100.0).round() as u32;
    InjectionPlan {
        strategy: InjectionStrategy::Chunked,
        allowance,
        document_tokens,
        budget,
        explanation: format!(
            "This document is about {document_tokens} tokens and only {allowance} were available, so roughly \
             the first {share}% of it was included. The rest is not in this turn."
        ),
    }
}

/// Truncates text to an allowance, on a character boundary.
///
/// Used only for the `Chunked` strategy when no retrieval index is available
/// for the document. The caller must say in the prompt that this happened —
/// text that stops mid-document with nothing marking it is how a model comes to
/// believe it has read a whole page.
pub fn take_tokens(text: &str, allowance: u32) -> String {
    let budget_chars = allowance as usize * CHARS_PER_TOKEN;
    if text.chars().count() <= budget_chars {
        return text.to_string();
    }
    text.chars().take(budget_chars).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_small_document_goes_in_whole() {
        let p = plan(1_000, 2_000, 32_000);
        assert_eq!(p.strategy, InjectionStrategy::Full);
    }

    #[test]
    fn a_large_document_is_chunked() {
        let p = plan(40_000, 4_000, 32_000);
        assert_eq!(p.strategy, InjectionStrategy::Chunked);
        assert_eq!(p.allowance, 14_000, "(32000 - 4000) * 0.5");
    }

    /// The boundary that must match `ocr-budget.test.ts` exactly. If these two
    /// ever disagree, the screen is describing a decision the prompt did not
    /// make.
    #[test]
    fn the_boundary_is_half_the_free_budget() {
        let free = 32_000 - 2_000;
        let exact = (f64::from(free) * FULL_INCLUSION_SHARE) as u32;
        assert_eq!(plan(exact, 2_000, 32_000).strategy, InjectionStrategy::Full);
        assert_eq!(
            plan(exact + 1, 2_000, 32_000).strategy,
            InjectionStrategy::Chunked
        );
    }

    #[test]
    fn a_document_never_takes_more_than_half_of_what_is_free() {
        for document in [1_000u32, 10_000, 100_000, 1_000_000] {
            let p = plan(document, 8_000, 32_000);
            assert!(
                p.allowance <= (32_000 - 8_000) / 2,
                "{document} was allowed {} of a 24000 free budget",
                p.allowance
            );
        }
    }

    #[test]
    fn a_nearly_full_turn_admits_nothing_rather_than_a_sliver() {
        let p = plan(50_000, 31_400, 32_000);
        assert_eq!(p.strategy, InjectionStrategy::ReferenceOnly);
        assert_eq!(p.allowance, 0);
        assert!(p.explanation.contains("no room"), "{}", p.explanation);
    }

    #[test]
    fn an_unknown_window_includes_everything_and_says_so() {
        let p = plan(90_000, 0, 0);
        assert_eq!(p.strategy, InjectionStrategy::Full);
        assert!(p.explanation.contains("not known"), "{}", p.explanation);
    }

    #[test]
    fn an_over_committed_turn_has_no_budget_rather_than_a_negative_one() {
        // `saturating_sub`, not a wrapping subtraction — which on u32 would
        // produce a budget of four billion and admit the document whole.
        let p = plan(500, 40_000, 32_000);
        assert_eq!(p.budget, 0);
        assert_eq!(p.strategy, InjectionStrategy::ReferenceOnly);
    }

    #[test]
    fn an_empty_document_estimates_to_zero_not_one() {
        assert_eq!(estimate_tokens(""), 0);
        assert_eq!(estimate_tokens("a"), 1, "a non-empty document is never zero");
    }

    #[test]
    fn truncation_respects_character_boundaries() {
        // Multi-byte input must not be cut mid-character; slicing by bytes
        // would panic on the first non-ASCII page the OCR model produces.
        let text = "नमस्ते दुनिया".repeat(50);
        let taken = take_tokens(&text, 4);
        assert_eq!(taken.chars().count(), 16);
    }

    #[test]
    fn truncation_leaves_short_text_alone() {
        assert_eq!(take_tokens("short", 100), "short");
    }
}
